### PR TITLE
feat: add scope for preheating job

### PR DESCRIPTION
**Pull request by gaius-qi**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by hyy0322 — DISMISSED** [review]

File: scheduler/job/job.go [hunk 1]
```diff
@@ -207,6 +209,52 @@ func (j *job) preheat(ctx context.Context, data string) error {
 	return nil
 }
 
+// preheatAllPeers preheats job by all peers, only suoported by v2 protocol.
+func (j *job) preheatAllPeers(ctx context.Context, req *internaljob.PreheatRequest) error {
+	eg, _ := errgroup.WithContext(ctx)
+	for _, host := range j.resource.HostManager().LoadAll() {
+		target := fmt.Sprintf("%s:%d", host.IP, host.Port)
+		eg.Go(func() error {
```
Comment (chlins) [inline scheduler/job/job.go:217]: Shouldn't we combine this with `batchSize` to limit the number of simultaneous tasks?

**Review by yxxhero — APPROVED** [review]

**Comment by yxxhero** [comment]
@gaius-qi ci

File: scheduler/job/job.go [hunk 2]
```diff
@@ -161,54 +158,197 @@ func (j *job) Serve() {
...
+// preheatAllPeers preheats job by all peers, only suoported by v2 protocol. Scheduler will trigger all peers to download task.
+// If all the peer download task failed, return error. If some of the peer download task failed, return success tasks and failure tasks.
+// Notify the client that the preheat is successful.
+func (j *job) preheatAllPeers(ctx context.Context, taskID string, req *internaljob.PreheatRequest, log *logger.SugaredLoggerOnWith) (*internaljob.PreheatResponse, error) {
+	var (
+		successTasks []*internaljob.PreheatSuccessTask
+		failureTasks []*internaljob.PreheatFailureTask
+	)
+
+	sem := semaphore.NewWeighted(req.ConcurrentCount)
```
Comment (chlins) [inline scheduler/job/job.go:244]: For controlling the concurrency, can we just use the `errgroup.SetLimit(n)`?

Comment (gaius-qi) [reply scheduler/job/job.go]: Fix 😊

File: internal/job/job.go [hunk 3]
```diff
@@ -179,6 +228,19 @@ func MarshalRequest(v any) ([]machineryv1tasks.Arg, error) {
 	}}, nil
 }
 
+func UnmarshalTaskResult(data interface{}, v any) error {
```
Comment (yxxhero) [inline internal/job/job.go:231]: interface{} same as any?

File: internal/job/job.go [hunk 4]
```diff
@@ -179,6 +228,19 @@ func MarshalRequest(v any) ([]machineryv1tasks.Arg, error) {
...
+	s, ok := data.(string)
+	if !ok {
+		return errors.New("invalid task result")
+	}
+
+	if err := json.Unmarshal([]byte(s), v); err != nil {
```
Comment (yxxhero) [inline internal/job/job.go:237]: return json.Unmarshal([]byte(s), v)

**Comment by yxxhero** [comment]
@gaius-qi could we add some e2e tests?

**Review by hhhhsdxxxx — APPROVED** [review]

**Review by chlins — APPROVED** [review]
lgtm

**Pull request merged**